File: src/com/brianc/layout/BlockBox.java
package com.brianc.layout;

import java.util.Optional;
import java.util.stream.Stream;

import com.brianc.css.KeywordValue;
import com.brianc.css.LengthValue;
import com.brianc.css.Unit;
import com.brianc.css.Value;
import com.brianc.graphics.Renderer;
import com.brianc.style.StyledNode;

public class BlockBox extends LayoutBox implements StyledLayoutBox {
	private final StyledNode styledNode;
	
	public BlockBox(final StyledNode styledNode) {
		super();
		this.styledNode = styledNode;
	}

	@Override
	public BoxType getType() {
		return BoxType.BLOCK_NODE;
	}

	@Override
	void layout(Dimensions containingBlock, Renderer renderBackend) {
		calculateBlockWidth(containingBlock);
		calculateBlockPosition(containingBlock);
		layoutBlockChildren(renderBackend);
		calculateBlockHeight();
	}

	private void calculateBlockHeight() {
		Optional<Value> heightValue = styledNode.value("height");
		if (heightValue.isPresent() && heightValue.get() instanceof LengthValue) {
			dimensions.content.height = heightValue.get().toPx();
		}
	}

	private void layoutBlockChildren(Renderer renderBackend) {
		for (LayoutBox child : children) {
			child.layout(dimensions, renderBackend);

			dimensions.content.height += child.dimensions.marginBox().height;
		}
	}

	private void calculateBlockPosition(Dimensions containingBlock) {
		Value zero = new LengthValue(0, Unit.PX);

		dimensions.margin.top = styledNode.lookup("margin-top", "margin", zero).toPx();
		dimensions.margin.bottom = styledNode.lookup("margin-bottom", "margin", zero).toPx();

		dimensions.margin.top = styledNode.lookup("border-left", "border", zero).toPx();
		dimensions.margin.bottom = styledNode.lookup("border-right", "border", zero).toPx();

		dimensions.padding.top = styledNode.lookup("padding-left", "padding", zero).toPx();
		dimensions.padding.bottom = styledNode.lookup("padding-right", "padding", zero).toPx();

		dimensions.content.x = containingBlock.content.x + dimensions.margin.left
				+ dimensions.border.left + dimensions.padding.left;

		dimensions.content.y = containingBlock.content.height + containingBlock.content.y
				+ dimensions.margin.top + dimensions.border.top + dimensions.padding.left;

	}

	private void calculateBlockWidth(Dimensions containingBlock) {
		Value auto = new KeywordValue("auto");
		Value width = styledNode.value("width").orElse(auto);
		Value zero = new LengthValue(0, Unit.PX);

		Value marginLeft = styledNode.lookup("margin-left", "margin", zero);
		Value marginRight = styledNode.lookup("margin-right", "margin", zero);

		Value borderLeft = styledNode.lookup("border-left-width", "border-width", zero);
		Value borderRight = styledNode.lookup("border-right-width", "border-width", zero);

		Value paddingLeft = styledNode.lookup("padding-left", "padding", zero);
		Value paddingRight = styledNode.lookup("padding-right", "padding", zero);

		float total = (float) Stream.of(marginLeft, marginRight, borderLeft, borderRight,
				paddingLeft, paddingRight, width).mapToDouble(v -> v.toPx()).sum();

		if (!width.equals(auto) && total > containingBlock.content.width) {
			if (marginLeft.equals(auto)) {
				marginLeft = zero;
			}

			if (marginRight.equals(auto)) {
				marginRight = zero;
			}
		}

		float underflow = containingBlock.content.width - total;

		boolean widthIsAuto = width.equals(auto);
		boolean leftIsAuto = marginLeft.equals(auto);
		boolean rightIsAuto = marginRight.equals(auto);

		if (!widthIsAuto) {
			if (!leftIsAuto && rightIsAuto) {
				marginRight = new LengthValue(underflow, Unit.PX);
			} else if (leftIsAuto && !rightIsAuto) {
				marginLeft = new LengthValue(underflow, Unit.PX);
			} else if (!leftIsAuto && !rightIsAuto) {
				marginRight = new LengthValue(marginRight.toPx() + underflow, Unit.PX);
			} else {
				marginLeft = new LengthValue(underflow / 2, Unit.PX);
				marginRight = new LengthValue(underflow / 2, Unit.PX);
			}
		} else {
			if (marginLeft.equals(auto)) {
				marginLeft = zero;
			}
			
			if (marginRight.equals(auto)) {
				marginRight = zero;
			}

			if (underflow >= 0) {
				width = new LengthValue(underflow, Unit.PX);
			} else {
				width = zero;
				marginRight = new LengthValue(marginRight.toPx() + underflow, Unit.PX);
			}
		}

		dimensions.content.width = width.toPx();

		dimensions.padding.left = paddingLeft.toPx();
		dimensions.padding.right = paddingRight.toPx();

		dimensions.border.left = borderLeft.toPx();
		dimensions.border.right = borderRight.toPx();

		dimensions.margin.left = marginLeft.toPx();
		dimensions.margin.right = marginRight.toPx();
	}

	@Override
	public StyledNode getStyledNode() {
		return styledNode;
	}
}

File: src/com/brianc/layout/BlockBox.kt
package com.brianc.layout

import com.brianc.css.KeywordValue
import com.brianc.css.LengthValue
import com.brianc.css.Unit
import com.brianc.css.Value
import com.brianc.graphics.Renderer
import com.brianc.style.StyledNode

class BlockBox(override val children: MutableList<LayoutBox> = mutableListOf(),
               override val dimensions: Dimensions = Dimensions(),
               override val styledNode: StyledNode) : StyledLayoutBox {
    override val inlineContainer: LayoutBox
        get() = if (children.all { it is InlineBox }) {
            this
        } else {
            children.add(AnonymousBlockBox())
            children.last()
        }

    override fun layout(containingBlock: Dimensions, renderBackend: Renderer) {
        calculateBlockWidth(containingBlock)
        calculateBlockPosition(containingBlock)
        layoutBlockChildren(renderBackend)
        calculateBlockHeight()
    }

    private fun calculateBlockHeight() {
        val heightValue = styledNode.values["height"]
        if (heightValue is LengthValue) {
            dimensions.content.height = heightValue.toPx()
        }
    }

    private fun layoutBlockChildren(renderBackend: Renderer) {
        for (child in children) {
            child.layout(dimensions, renderBackend)

            dimensions.content.height += child.dimensions.marginBox().height
        }
    }

    private fun calculateBlockPosition(containingBlock: Dimensions) {
        val zero = LengthValue(0f, Unit.PX)

        dimensions.margin.top =
                styledNode.lookup("margin-top", "margin", zero).toPx()
        dimensions.margin.bottom =
                styledNode.lookup("margin-bottom", "margin", zero).toPx()

        dimensions.border.top =
                styledNode.lookup("border-left", "border", zero).toPx()
        dimensions.border.bottom =
                styledNode.lookup("border-right", "border", zero).toPx()

        dimensions.padding.top =
                styledNode.lookup("padding-left", "padding", zero).toPx()
        dimensions.padding.bottom =
                styledNode.lookup("padding-right", "padding", zero).toPx()

        dimensions.content.x = containingBlock.content.x + dimensions.margin.left + dimensions.border.left + dimensions.padding.left

        dimensions.content.y = containingBlock.content.height + containingBlock.content.y + dimensions.margin.top + dimensions.border.top + dimensions.padding.left
    }

    private fun calculateBlockWidth(containingBlock: Dimensions) {
        val auto = KeywordValue("auto")
        val width = styledNode.values["width"] ?: auto
        val zero = LengthValue(0f, Unit.PX)

        var marginLeft: Value = styledNode.lookup("margin-left", "margin", zero)
        var marginRight: Value = styledNode.lookup("margin-right", "margin", zero)

        val borderLeft = styledNode.lookup("border-left-width", "border-width", zero)
        val borderRight = styledNode.lookup("border-right-width", "border-width", zero)

        val paddingLeft = styledNode.lookup("padding-left", "padding", zero)
        val paddingRight = styledNode.lookup("padding-right", "padding", zero)

        val total = listOf(marginLeft, marginRight, borderLeft, borderRight,
                paddingLeft, paddingRight, width)
                .filterIsInstance<LengthValue>()
                .map(LengthValue::toPx).sum()

        if (width != auto && total > containingBlock.content.width) {
            if (marginLeft == auto) {
                marginLeft = zero
            }

            if (marginRight == auto) {
                marginRight = zero
            }
        }

        val underflow = containingBlock.content.width - total

        val widthIsAuto = width == auto
        val leftIsAuto = marginLeft == auto
        val rightIsAuto = marginRight == auto

        var leftMarginPx = 0f
        var rightMarginPx = 0f
        var widthPx = 0f

        if (!widthIsAuto) {
            widthPx = (width as LengthValue).toPx()
            when {
                (!leftIsAuto && rightIsAuto) -> {
                    rightMarginPx = underflow
                }
                (leftIsAuto && !rightIsAuto) -> {
                    leftMarginPx = underflow
                }
                (!leftIsAuto && !rightIsAuto) -> {
                    rightMarginPx = (marginRight as LengthValue).toPx() + underflow
                }
                else -> {
                    leftMarginPx = underflow / 2
                    rightMarginPx = underflow / 2
                }
            }
        } else {
            if (marginLeft == auto) {
                leftMarginPx = 0f
            }

            if (marginRight == auto) {
                rightMarginPx = 0f
            }

            if (underflow >= 0) {
                widthPx = underflow
            } else {
                rightMarginPx = (marginRight as LengthValue).toPx() + underflow
            }
        }

        dimensions.content.width = widthPx

        dimensions.padding.left = paddingLeft.toPx()
        dimensions.padding.right = paddingRight.toPx()

        dimensions.border.left = borderLeft.toPx()
        dimensions.border.right = borderRight.toPx()

        dimensions.margin.left = rightMarginPx
        dimensions.margin.right = leftMarginPx
    }
}
